pkg/pubsub/broker/nats.go

```diff
@@ -0,0 +1,157 @@
+	// js represents the jetstream API on top of the NATS connection
+	js jetstream.JetStream
+	// consumers stores the list of consumers used for this connection.
+	// this is only available when jetstream is enabled.
+	//
+	// the key is a combination of stream name + consumer name (e.g. trace + trace-consumer = "trace-trace-consumer")
+	consumers map[string]jetstream.Consumer
+}
+
+func NewNATSConnector(ctx context.Context, opts NatsConnOpt) (*NatsConnector, error) {
+	host, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving hostname: %w", err)
+	}
+	name := fmt.Sprintf("%s-%s", opts.Name, host)
+
+	urls := "nats://127.0.0.1:4222"
+	if len(opts.URLS) > 0 {
+		urls = opts.URLS
+	}
+
+	wg := sync.WaitGroup{}
+
+	// merge provided options
+	nopts := []nats.Option{
+		nats.Name(name),
+		nats.ClosedHandler(func(c *nats.Conn) {
+			wg.Done()
+		}),
+	}
+	nopts = append(nopts, opts.Opts...)
+
+	conn, err := nats.Connect(urls, nopts...)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to NATS: %w", err)
+	}
+	wg.Add(1)
+
+	logger.StdlibLogger(ctx).Info("established connection with NATS server",
+		"urls", opts.URLS,
+		"name", name,
+	)
+
+	c := &NatsConnector{
+		connWg: &wg,
+		conn:   conn,
+	}
+
+	if opts.JetStream {
+		// NOTE: should there be some default jetstream options?
+		js, err := jetstream.New(conn)
+		if err != nil {
+			return nil, fmt.Errorf("error initializing jetstream API: %w", err)
+		}
+		c.js = js
+
+		// initialize the consumers
+		if opts.Consumers != nil {
+			c.consumers = map[string]jetstream.Consumer{}
+
+			for stream, consumers := range opts.Consumers {
+				for _, consumer := range consumers {
+					cons, err := js.Consumer(ctx, stream, consumer)
+					if err != nil {
+						return nil, fmt.Errorf("error initializing consumer: %w", err)
+					}
+
+					key := fmt.Sprintf("%s-%s", stream, consumer)
+					c.consumers[key] = cons
+				}
+			}
+		}
+	}
+
+	return c, nil
+}
+
+// Publish submits the provided data. It'll use jetstream if it's available, or default
+// to the core NATS connection.
+func (c *NatsConnector) Publish(ctx context.Context, sub string, data []byte) error {
+	if c.js != nil {
+		_, err := c.js.Publish(ctx, sub, data)
+		return err
+	}
+	return c.conn.Publish(sub, data)
+}
+
+// Consumer returns the specified consumer based on stream name and consumer name, if it's initialized
+func (c *NatsConnector) Consumer(ctx context.Context, stream, consumer string) (jetstream.Consumer, error) {
+	if c.consumers == nil {
+		return nil, fmt.Errorf("consumers are not initialized")
+	}
+
+	key := fmt.Sprintf("%s-%s", stream, consumer)
+	if cons, ok := c.consumers[key]; ok {
+		return cons, nil
+	}
+
+	return nil, fmt.Errorf("no consumer named '%s' initialized for stream: '%s'", consumer, stream)
+}
+
+// Shutdown drains the connection to flush data in buffer before fully closing the
+// connection
+func (c *NatsConnector) Shutdown(ctx context.Context) error {
```

Comment:
**question**: Do we have to do anything to clean up the consumers? Like `DeleteConsumer`?

Reply:
No, it shouldn't be clean up.
A consumer is a nats resource, created on the nats server side. we merely authenticate as a consumer to pull messages to work on.